openapi-style-guide.md

```diff
@@ -0,0 +1,208 @@
+
+Further, a put or patch that supports both create and update should have "Create" and "Update" in the verb.
+
+### Request body
+
+The request body for a patch operation should be `application/merge-patch+json` content type.
+
+### Response body
+
+All success responses except 202 and 204 should define a response body.
+
+Responses for status codes 202 and 204 should have no response body.
+
+A delete operation should have a 204 response.
+
+For a path with a "create" operation (put or patch that returns 201),
+the 200 response of get, put, and patch, if present, should have the same response body schema
+as the create operation 201 response.
+
+### Support for pagination
+
+To support pagination:
+- The operation should have the `x-ms-pageable` annotation
+- The operation response should contain a top-level `value` property of type array and required
+- The operation response should contain a top-level `nextLink` property of type string and optional
+- If the operation has a `skip` parameter, it must be an integer and optional
+- If the operation has a `top` parameter, it must be an integer, optional, and have a documented default and maximum value
+- If the operation has a `maxpagesize` parameter, it must be an integer, optional, and have a documented default and maximum value
+
+### Error response
+
+All operations should have a "default" (error) response with a response body that conforms to the Azure API Guidelines.
+
+All `4xx` and `5xx` responses should specify `x-ms-error-response: true` except for `404` response of HEAD operation.
+
+Every error response should include the `x-ms-error-code` response header in the `headers` of the response.
+
+Example:
+```
+      "400": {
+        "description": "Bad Request",
+        "headers": {
+          "x-ms-error-code": {
+            "type": "string",
+            "description": "The error code."
+          }
+        },
+        "schema": {
+          "$ref": "#/definitions/ErrorResponse"
+        },
+        "x-ms-error-response": true
+      },
+```
+
+### Response headers
+
+A 202 response should include the `Operation-Location` response header in the `headers` of the response.
+
+Example:
+```
+    "responses": {
+      "202": {
+        "description": "The service has accepted the request and will start processing later.",
+        "headers": {
+          "Operation-Location": {
+            "description": "URL to query for status of the operation.",
+            "type": "string"
+          }
+        }
+      },
+```
+
+<!-- --------------------------------------------------------------- -->
+
+## Parameters
+
+### Parameter names
+
+All parameter names for an operation -- including parameters defined at the path level -- should be case-insensitive unique.
+
+### Descriptions
+
+Every parameter should have a description.
+
+### Format
+
+Format must be one of the values [defined by OpenAPI][openapi-data-types] and recognized by the Azure tooling.
```

Comment:
I don't fully agree with this one.  We use more than the official spec (things like `"format": "date-time-rfc1123"` come to mind).  I'm not sure there's an official list recognized by AutoRest, but we should document that and tell people those are the allowed formats.

Reply:
I modeled this rule after the [ARM validation rule R2003](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#r2003).  That rule points to [this page](https://github.com/Azure/autorest/blob/81d4d31d06637f4f9ef042d7f2ec64cfea29892f/docs/developer/validation-rules/valid-formats.md), which then points [here](https://github.com/Azure/autorest/blob/81d4d31d06637f4f9ef042d7f2ec64cfea29892f/docs/developer/guide/defining-clients-swagger.md#data-types) for a list of valid formats.  Unfortunately, this last link really don't give clear guidance on valid formats, so I resorted to just linking to the OpenAPI 2.0 spec.

So I did look for an official list recognized by Autorest and came up empty.  But I can make another run at that.  I can also look at the formats we already use in our API docs as another data point for this.

Reply:
That's a generation or two old.  Let's not block your work on this, but we should follow up with the folks who own ModelerFour and make them document the exact list somewhere.  I think it's  https://github.com/Azure/autorest/blob/main/packages/libs/openapi/src/formats.ts though.

Reply:
I'm going to change the "and" to "or", since we definitely accept formats that are not defined in the OpenAPI spec.